db/bigtable_eth1.go

```diff
@@ -550,40 +538,54 @@ func (bigtable *Bigtable) GetBlocksDescending(start, limit uint64) ([]*types.Eth
 	})
 	defer tmr.Stop()
 
-	if start < 1 || limit < 1 || limit > start {
-		return nil, fmt.Errorf("invalid block range provided (start: %v, limit: %v)", start, limit)
+	if limit == 0 {
+		return nil, fmt.Errorf("error limit is set to 0 which would not fetch any blocks")
 	}
 
-	startPadded := reversedPaddedBlockNumber(start)
-	endPadded := reversedPaddedBlockNumber(start - limit)
+	// clamp limit
+	if limit > start+1 {
+		limit = start + 1
+	}
 
-	// logger.Info(start, start-limit)
-	// logger.Info(startPadded, " ", endPadded)
 	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(time.Second*30))
 	defer cancel()
 
-	startKey := fmt.Sprintf("%s:B:%s", bigtable.chainId, startPadded)
-	endKey := fmt.Sprintf("%s:B:%s", bigtable.chainId, endPadded)
+	blocks := make([]*types.Eth1BlockIndexed, 0, limit)
 
-	rowRange := gcp_bigtable.NewRange(startKey, endKey) //gcp_bigtable.PrefixRange("1:1000000000")
+	rowHandler := getBlockHandler(&blocks)
+	rowFilter := gcp_bigtable.RowFilter(gcp_bigtable.ColumnFilter("d"))
 
-	if limit >= start { // handle retrieval of the first blocks
-		rowRange = gcp_bigtable.InfiniteRange(startKey)
-	}
+	if start > 0 {
+		startKey := fmt.Sprintf("%s:B:%s", bigtable.chainId, reversedPaddedBlockNumber(start))
+		endKey := fmt.Sprintf("%s:B:%s", bigtable.chainId, reversedPaddedBlockNumber(start-limit))
+		if limit >= start {
+			// don't include the broken block 0 in the range
```

Comment:
Block 0 is padded incorrectly. When its row key is used for a range, this means that (iirc) no block is returned at all as the start range is after the end range. Broken blocks cannot be returned, if something goes wrong its either because of no blocks at all or because of "wrong" blocks.

If you want to dive deeper, let's discuss this in the office.

Regarding `limit >= start` versus `limit > start`: If you use `1` for both values (meaning "please get me only block 1"), we want that check to be triggered as otherwise `endKey` would have the row key of block 0 because of `start-limit`.

I have improved this code with another commit though to harmonize the logic with `GetFullBlocksDescending`: 79aab64d428bfe042015d4cdb27d6e413afc72ba

ptal